qutebrowser/keyinput/basekeyparser.py

```diff
@@ -29,6 +31,79 @@
 from qutebrowser.keyinput import keyutils
 
 
+@attr.s(slots=True)
+class BindingTrie:
+
+    """Helper class for key parser. Represents a set of bindings.
+
+    Except for the root item, there is no children BindingTrie with no bound
+    commands.
+
+    This class works like a Mapping[keyutils.KeySequence], but with matches
+    method added. Note that some unused methods may not be implemented. It may
+    be a good idea to make this function inherit from
+    collections.abc.MutableMapping.
+
+    Attributes:
+        child: A map. Keys of this map can be get from the KeyInfo.to_int
+               method.
+        command: Command associated with the root trie node.
+    """
+
+    child = attr.ib(type=Mapping[int, 'BindingTrie'], factory=dict)
```

Comment:
Why use attrs here at all? This doesn't seem like a class which benefits from using attrs much.

Reply:
Main benefit is being able to slot the class easily for correctness and memory usage, but I think it's easier to manually slot things than deal with attrs :)

Reply:
FWIW I don't think using `__slots__` would really show any usable benefit here. It's not that we'd have [thousands](http://tech.oyster.com/save-ram-with-python-slots/) of instances of this, right?

Reply:
FWIW, we already have ~323 with the default config, it's I wouldn't think it's unreasonable to see this go up to thousands with deeper binding configurations (my personal binding without the per-domain ones goes up to 400 already, and I can see adding a lot more). We also get stricter variable handling as well with slotting :).

(these kinds of things are a long-tail, and while it's very hard to measure it directly, it will provide a improvement when used consistently across the entire program. Of course, we don't do that, but we have to start somewhere.)

Reply:
Even if they only provide 0.01% perf improvement, they don't hurt code readability either, so I don't see any reason why they should not be added.

Reply:
@jgkamat "deal with" attrs? Is attrs a problem? ...

Reply:
After using attrs on several small size projects of my own, I can confidently say it creates major complications and limits what you can do with classes (which can only be fixed with messy hacks or by not using attrs). While attrs is slightly nicer for extremely simple classes, it creates a maintenance/readability burden for future modification. I don't personally mind though, as I'm not dealing with maintaining this and I'm planning to avoid related code [as much as possible](http://www.github.com/qutebrowser/qutebrowser/pull/4667#discussion_r276873739) (and because this is one of those 'simple cases', for now at least).

- http://www.github.com/python-attrs/attrs/issues/120
- http://www.github.com/python-attrs/attrs/issues/491
- http://www.github.com/python-attrs/attrs/issues/38#issuecomment-313113090 (worked around, but not in a way that fixes the actual original issue)
- http://www.github.com/python-attrs/attrs/issues/448
- http://www.github.com/python-attrs/attrs/issues/342
- http://www.github.com/python-attrs/attrs/issues/393
- http://www.github.com/python-attrs/attrs/pull/172

> Update: That fails on py35-pyqt571. How annoying.

You can change the classname to a string for compatibility. [example](http://www.github.com/qutebrowser/qutebrowser/pull/4731/files#diff-6a35a1a9b55a6e095c37c8d1b34c5962R42)